subprojects/core/src/main/java/org/gradle/execution/taskgraph/DefaultTaskExecutionPlan.java

```diff
@@ -480,7 +480,11 @@ public void getNodeValues(TaskInfo node, Collection<? super Void> values, Collec
             }
         });
         graphWalker.add(entryTasks);
-        final List<TaskInfo> firstCycle = new ArrayList<TaskInfo>(graphWalker.findCycles().get(0));
+        List<Set<TaskInfo>> cycles = graphWalker.findCycles();
+        if (cycles.isEmpty()) {
```

Comment:
It doesn't make sense to me that we'd call `onOrderingCycle` when there actually is no cycle. That looks to me more like the algorithm that calls this method is broken. I.e. there is https://github.com/fisakov/gradle/blob/ed890eb39ef876aae957b666725702e1b63cf0c0/subprojects/core/src/main/java/org/gradle/execution/taskgraph/DefaultTaskExecutionPlan.java#L308 which looks like it should take care of this already.

Reply:
I agree, this algorithm is broken. For example, in this particular scenario, it starts with 'a' and pushes 'c' and 'b' as its "shouldRunAfter" successors, but when it actually arrives at 'c' it does so after having visited 'd' (added to the queue when visiting 'b' in the meantime), and the latter one doesn't have 'c' in its `shouldSuccessors`, so the edge doesn't get recorded by `recordEdgeIfArrivedViaShouldRunAfter()`. Thus we fall into the second branch of `if (visitingNodes.containsEntry(successor, currentSegment))` and therefore a cycle is found. 

What do you suggest? In my opinion, this code needs to be rewritten in order to be verifiable or at least readable. At this stage it's nearly impossible to predict what it actually does.

Reply:
Then I'd suggest refactoring this code before you attempt the fix. There are plenty of tests verifying that task ordering works, so you have a good safety net.